File: Difficulty: Medium/Longest Subarray with Majority Greater than K/longest-subarray-with-majority-greater-than-k.cpp
class Solution {
  public:
    int longestSubarray(vector<int> &arr, int k) {
        // Code here
        int n=arr.size();
       
        map<long,int>mp;
        int ans=0;
        int count=0;
        for(int i=0;i<n;i++)
        {
            if(arr[i]>k)
            {
                ++count;
            }
            if(arr[i]<=k)
            {
                --count;
            }
            if(count>0)
            {
               ans=max(ans,i+1); 
            }
            else
            {
                auto x=mp.lower_bound(count);
                if(x!=mp.begin())
                
                {
                    --x;
                }
                if(x->first<count)
                {
                    ans=max(ans,i-x->second);
                }
            }
            if(!mp.count(count))
            {
                mp[count]=i;
            }
        }
        return ans;
        
    }
};

File: Difficulty: Medium/Longest Subarray with Majority Greater than K/longest-subarray-with-majority-greater-than-k.java
class Solution {
    public int longestSubarray(int[] arr, int k) {
        // Code Here
        HashMap<Integer,Integer>mp=new HashMap<>();
        int n=arr.length;
        int ans=0;
        int count=0;
        for(int i=0;i<n;i++)
        {
            if(arr[i]>k)
            {
                ++ans;
            }
            else
            {
                --ans;
            }
            if(ans>0)
            {
                count=Math.max(count,i+1);
            }
            else if(mp.containsKey(ans-1))
            {
                count=Math.max(count,(i-(mp.get(ans-1))));
            }
            if(!mp.containsKey(ans))
            {
                mp.put(ans,i);
            }
               
        }
        return count;
        
    }
}